contracts/airdrops.js

```diff
@@ -0,0 +1,244 @@
+/* eslint-disable no-await-in-loop */
+/* global actions, api */
+
+// eslint-disable-next-line no-template-curly-in-string
+const UTILITY_TOKEN_SYMBOL = "'${CONSTANTS.UTILITY_TOKEN_SYMBOL}$'";
+const CONTRACT_NAME = 'airdrops';
+
+actions.createSSC = async () => {
+  const tableExists = await api.db.tableExists('pendingAirdrops');
+  if (tableExists === false) {
+    await api.db.createTable('pendingAirdrops', ['airdropId', 'symbol']);
+    await api.db.createTable('params');
+
+    const params = {};
+    params.feePerTransaction = '0.1';
+    params.maxTransactionsPerBlock = 50;
+    params.maxAirdropsPerBlock = 1;
+    await api.db.insert('params', params);
+  }
+};
+
+actions.updateParams = async (payload) => {
+  if (api.assert(api.sender === api.owner, 'not authorized')) {
+    const {
+      feePerTransaction,
+      maxTransactionsPerBlock,
+      maxAirdropsPerBlock,
+    } = payload;
+
+    const params = await api.db.findOne('params', {});
+
+    if (feePerTransaction) {
+      if (!api.assert(typeof feePerTransaction === 'string' && !api.BigNumber(feePerTransaction).isNaN() && api.BigNumber(feePerTransaction).gte(0), 'invalid feePerTransaction')) return;
+      params.feePerTransaction = feePerTransaction;
+    }
+    if (maxTransactionsPerBlock) {
+      if (!api.assert(Number.isInteger(maxTransactionsPerBlock) && maxTransactionsPerBlock >= 1, 'invalid maxTransactionsPerBlock')) return;
+      params.maxTransactionsPerBlock = maxTransactionsPerBlock;
+    }
+    if (maxAirdropsPerBlock) {
+      if (!api.assert(Number.isInteger(maxAirdropsPerBlock) && maxAirdropsPerBlock >= 1, 'invalid maxAirdropsPerBlock')) return;
+      params.maxAirdropsPerBlock = maxAirdropsPerBlock;
+    }
+
+    await api.db.update('params', params);
+  }
+};
+
+const parseAirdrop = async (list, precision) => {
+  const params = await api.db.findOne('params', {});
+  const airdrop = {};
+  airdrop.list = [];
+  airdrop.fee = '0';
+  airdrop.quantity = '0';
+  airdrop.isValid = false;
+
+  // loop through list and validate
+  for (let i = 0; i < list.length; i += 1) {
+    const { 0: to, 1: quantity } = list[i];
+
+    if (to && api.isValidAccountName(to)
+      && quantity && !api.BigNumber(quantity).isNaN()
+      && api.BigNumber(quantity).gt(0) && api.BigNumber(quantity).dp() <= precision) {
```

Comment:
I didn't get that?

Reply:
Since you will report it as invalid if any entry is not good, you can exit as soon as you detect a problem. Also optional: report entry that failed (can emit a log or just api assert)

Reply:
ah, yes. it should not loop through whole list after finding error in the first object.